crypto/default/src/main/java/org/keycloak/crypto/def/DefaultCryptoProvider.java

```diff
@@ -0,0 +1,38 @@
+package org.keycloak.crypto.def;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.keycloak.common.crypto.CryptoProvider;
+import org.keycloak.common.crypto.CryptoProviderTypes;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class DefaultCryptoProvider implements CryptoProvider {
+
+    private Map<String, Supplier<?>> providers = new HashMap<>();
+
+    public DefaultCryptoProvider() {
+        providers.put(CryptoProviderTypes.BC_SECURITY_PROVIDER, BouncyCastleProvider::new);
+        providers.put(CryptoProviderTypes.AES_KEY_WRAP_ALGORITHM_PROVIDER, AesKeyWrapAlgorithmProvider::new);
+    }
+
+    @Override
+    public SecureRandom getSecureRandom() throws NoSuchAlgorithmException {
+        return SecureRandom.getInstance("SHA1PRNG");
+    }
+
+    @Override
+    public <T> T getCryptoUtility(Class<T> clazz, String typeId) {
```

Comment:
This stuff seems an additional layer of indirection that perhaps isn't really needed? Wouldn't it be cleaner to just have `CryptoProvider` have methods to directly obtain what's needed?

Reply:
That's what I was initially thinking, but looks harder to achieve due the reasons like:

- The CryptoProvider is on the keycloak-common, but some method arguments are available just on the `keycloak-core` (or other modules). For example interface `JWEAlgorithmProvider` itself is in the `keycloak-core` and it has 2 methods, which both has arguments classes also available only in the `keycloak-core` . So adding directly methods like `encodeAesKeyWrap` and `decodeAesKeyWrap` on the CryptoProvider doesn't simply work (without moving some more classes from keycloak-common to the keycloak-core and/or doing some more changes in the arguments etc. In general, harder to achieve and not so nice).
- Also there will be too much methods on the CryptoProvider, which IMO will make it rather bloated class.

Reply:
I didn't mean adding the methods directly, but rather specific methods to get specific providers.

Instead of:

`Provider provider = CryptoIntegration.getProvider().getCryptoUtility(Provider.class, CryptoProviderTypes.BC_SECURITY_PROVIDER)`

You'd have:

`Provider provider = CryptoIntegration.getProvider().getBCSecurityProvider()`

Reply:
@stianst This works if the returned class like `Provider` is available on the classpath of `keycloak-common`. It doesn't work if returned class like `JWEAlgorithmProvider` is itself not available on the classpath of `keycloak-common`. For that case, the alternative is to:

1) return `Object` as return type:
```
Object getAesKeyWrapAlgorithm();
```
And the caller would need to retype something like:
```
JWEAlgorithmProvider alg =  (JWEAlgorithmProvider) CryptoIntegration.getProvider().getAesKeyWrapAlgorithm();
```

2) or to avoid casting something like:
```
T getAesKeyWrapAlgorithm(Class<T> clazz);
```
And the caller do this:
```
JWEAlgorithmProvider alg =  CryptoIntegration.getProvider().getAesKeyWrapAlgorithm(JWEAlgorithmPRovider.class);
```

Which one do you prefer? Or any other idea?